Add PRD for DRA support (#1275)
kubernetes-sigs/kernel-module-management: docs/enhancements/0004-dra-support.md

# Adding DRA Support to KMM

| Field       | Value   |
|-------------|---------|
| Author(s)   | Tomer Newman |
| Date        | 2026-05-26 |

## 1. Problem Statement

KMM manages the lifecycle of out-of-tree kernel modules and their associated Device Plugins. Device Plugins use a static integer counting model for hardware resource advertisement — a mechanism Kubernetes is replacing with Dynamic Resource Allocation (DRA). As the ecosystem adopts DRA, KMM users who need modern hardware orchestration (structured parameters, resource claims, device classes) have no way to deploy DRA drivers through KMM. Without DRA support, KMM users must manage DRA driver DaemonSets and DeviceClass resources outside of KMM, losing the operator's lifecycle management, ordered upgrades, and integration with kernel module loading.

## 2. Goals and Non-Goals

### 2.1 Goals

> - KMM users can deploy a DRA kubelet plugin alongside a kernel module by specifying `spec.dra` on a Module resource.

Reviewer: ```suggestion
- KMM users can deploy a DRA driver alongside a kernel module by specifying `spec.dra` on a Module resource.
```

Author: Good catch, updated. Thanks!

- The KMM operator manages the full lifecycle of DRA DaemonSets (create, update, delete) and DeviceClass resources on behalf of the user.
- DRA modules participate in ordered upgrades with the same guarantees as Device Plugin modules.

> - The admission webhook prevents users from creating invalid DRA configurations (unsupported cluster version, missing or invalid driver name).

Reviewer: We should also specify that in KMM Module DRA Spec and DevicePlugin Spec are mutually exclusive

Author: Agreed — added mutual exclusion as an explicit goal. This is also enforced via FR-3 and the acceptance criteria.

Reviewer 2: This is already explained [here](https://github.com/kubernetes-sigs/kernel-module-management/pull/1275/changes#diff-e741503ca81e3ea56628c8eeff173ccb6355bbd4e7c423854846ba0592984d5aR36) which is the correct place to mention it IMO.

Author: Fair point — mutual exclusion is already captured in FR-3 and the acceptance criteria. Removing it from the goals to avoid redundancy.

### 2.3 Non-Goals

- Hub/multi-cluster DRA support — DRA operates in spoke mode only, matching DevicePlugin behavior.
- In-place migration from DevicePlugin to DRA — users delete the existing Module and create a new one.
- DRA-specific Prometheus metrics.

## 3. Requirements

### 3.1 Functional Requirements

#### CRD and API

> - **FR-1:** The Module CRD must include a `spec.dra` field that accepts a DRA plugin configuration: container spec, init container, service account, volumes, `driverName` (required, DNS subdomain), and `deviceClasses` (optional list of DeviceClass definitions).

Reviewer: ```suggestion
- **FR-1:** The Module CRD must include a `spec.dra` field that accepts a DRA  configuration: container spec, init container, service account, volumes, `driverName` (required, DNS subdomain), and `deviceClasses` (optional list of DeviceClass definitions).
```

Author: Updated, thanks.

- **FR-2:** DRA and DevicePlugin container/pod properties must share common base types (`PluginSpec`, `PluginContainerSpec`). `DevicePluginContainerSpec` must remain as a type alias for backward compatibility.
- **FR-3:** `spec.dra` and `spec.devicePlugin` must be mutually exclusive. The CRD must reject a Module that sets both fields.
- **FR-4:** `status.dra` must report the DRA DaemonSet's availability using the same `DaemonSetStatus` structure as `status.devicePlugin`.

#### Webhook Validation

> - **FR-5:** The admission webhook must reject creation of a Module with `spec.dra` when the cluster Kubernetes version is below 1.35.

Reviewer: ```suggestion
- **FR-5:** The admission webhook must reject creation of a Module with `spec.dra` when the cluster Kubernetes version is below 1.34.
```

Author: Fixed — updated all references from 1.35 to 1.34.

- **FR-10:** The DRA DaemonSet must target nodes where the kernel module is loaded (using the kernel-module-ready node label as its nodeSelector).

#### DeviceClass Management

- **FR-11:** The DRA controller must create, update, and delete Kubernetes DeviceClass resources based on `spec.dra.deviceClasses`.
- **FR-12:** DeviceClass ownership must be tracked via labels on the DeviceClass (Module name and namespace), since DeviceClass is cluster-scoped and Module is namespaced.
- **FR-13:** When a Module is deleted, all DeviceClasses managed by that Module must be deleted.

#### Ordered Upgrades

- **FR-14:** DRA must participate in KMM ordered upgrades with the same behavior as DevicePlugin: DRA-specific version labels on nodes, a DRA pod reconciler that labels nodes on pod readiness, extension of the label action table for DRA label transitions, and garbage collection of old-version DRA DaemonSets.

### 3.2 Non-Functional Requirements

- **NFR-1:** All CRD changes must be backward compatible. Existing Modules using `spec.devicePlugin` must continue to work without modification.

> - **NFR-2:** DRA features require Kubernetes >= 1.35. The webhook enforces this at admission time.

Reviewer: ```suggestion
- **NFR-2:** DRA features require Kubernetes >= 1.34. The webhook enforces this at admission time.
```

Author: Fixed — updated all references from 1.35 to 1.34.

## 4. Acceptance Criteria

- [ ] A Module with `spec.dra` set results in a running DRA DaemonSet on all nodes matching the module selector where the kernel module is loaded.
- [ ] The DRA DaemonSet mounts `/var/lib/kubelet/plugins/` and `/var/lib/kubelet/plugins_registry/` automatically.
- [ ] DeviceClass resources listed in `spec.dra.deviceClasses` are created in the cluster. Removing an entry from the list deletes the corresponding DeviceClass. Editing an entry updates the DeviceClass.

> - [ ] The webhook rejects a Module with `spec.dra` on a cluster running Kubernetes < 1.35.

Reviewer: ```suggestion
- [ ] The webhook rejects a Module with `spec.dra` on a cluster running Kubernetes < 1.34.
```

Author: Fixed — updated all references from 1.35 to 1.34.

- [ ] The webhook rejects a Module with `spec.dra` when `driverName` is empty or not a valid DNS subdomain.
- [ ] A Module cannot have both `spec.dra` and `spec.devicePlugin` set simultaneously.
- [ ] DRA DaemonSets participate in ordered upgrades: version labels are set on nodes, old-version DaemonSets are garbage-collected when no longer scheduled, and version transitions follow the same rules as DevicePlugin.
- [ ] Deleting a Module with `spec.dra` removes all associated DRA DaemonSets and DeviceClasses.
- [ ] Existing Modules with `spec.devicePlugin` (and no `spec.dra`) continue to work without any changes.

## 5. Assumptions

> - The Kubernetes DRA API (`k8s.io/api/resource/v1`) is stable and available on clusters running version 1.35+.

Reviewer: ```suggestion
- The Kubernetes DRA API (`k8s.io/api/resource/v1`) is stable and available on clusters running version 1.34+.
```

Author: Fixed — updated all references from 1.35 to 1.34.

- DRA kubelet plugins register via Unix sockets under `/var/lib/kubelet/plugins/` and discover through `/var/lib/kubelet/plugins_registry/`, following the standard Kubernetes plugin registration mechanism.

## 6. Dependencies

> - Kubernetes >= 1.35 with DRA support enabled.

Reviewer: ```suggestion
- Kubernetes >= 1.34 with DRA support enabled.
```

Author: Fixed — updated all references from 1.35 to 1.34.
